packages/use-optionally-controlled-state/test/index.test.tsx

```diff
@@ -102,3 +110,64 @@ it('throws when switching from controlled to uncontrolled mode', () => {
     /Can not change from controlled to uncontrolled mode./
   );
 });
+
+/**
+ * Type signature tests
+ */
+
+function TestTypes() {
+  const controlled = useOptionallyControlledState({
+    controlledValue: true
+  });
+  const uncontrolledWithInitialValue = useOptionallyControlledState({
+    initialValue: true
+  });
+  const uncontrolledWithoutInitialValue = useOptionallyControlledState({});
+
+  // Only used for type tests; mark the variables as used
+  // eslint-disable-next-line no-unused-expressions
+  [controlled, uncontrolledWithInitialValue, uncontrolledWithoutInitialValue];
+}
+
+// Expected return type: `[boolean, (value: boolean) => void]`
+function Controlled(opts: {controlledValue: boolean; initialValue?: boolean}) {
+  const [value, setValue] = useOptionallyControlledState(opts);
+
+  setValue(true);
+  return value.valueOf();
+}
+
+// Expected return type: `[boolean, (value: boolean) => void]`
+function UncontrolledWithInitialValue(opts: {
+  controlledValue?: boolean;
+  initialValue: boolean;
+}) {
+  const [value, setValue] = useOptionallyControlledState(opts);
+
+  setValue(true);
+  return value.valueOf();
+}
+
+// Expected return type: `[boolean | undefined, (value: boolean) => void]`
+function UncontrolledWithoutInitialValue(opts: {
+  controlledValue?: boolean;
+  initialValue?: boolean;
+}) {
+  // False positive, probably need to upgrade ESLint
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [value, setValue] = useOptionallyControlledState(opts);
+
+  setValue(true);
+
+  // @ts-expect-error This should be an error as the value can be undefined
+  return value.valueOf();
```

Comment:
This should be an error but isn't unfortunately.

Reply:
@menelaos Do you have an idea by chance?

Reply:
I just tried for quite a bit, but couldn't find a solution 😞 I would also suggest to proceed with the potential `undefined` for now.